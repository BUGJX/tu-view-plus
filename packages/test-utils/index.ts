export { default as defineGetter } from './define-getter';
