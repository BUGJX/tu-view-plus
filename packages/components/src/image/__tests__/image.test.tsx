import { mount } from '@vue/test-utils';
import { describe, expect, test, vi, beforeAll, afterAll } from 'vitest';
import Image from '../src/image.vue';
import ImagePreview from '../src/image-preview.vue';
import ImagePreviewGroup from '../src/image-preview-group.vue';
import getScale from '../utils/get-scale';

const imgSrc = 'http://127.0.0.1:9300/tu-view-plus/logo.svg';

async function getPreviewInstance() {
  const wrapper = await mount(ImagePreview, {
    props: {
      src: imgSrc,
      defaultVisible: true,
      renderToBody: false
    }
  });
  await wrapper.vm.onImgLoad();
  return wrapper;
}

describe('Image.vue', () => {
  beforeAll(() => {
    vi.useFakeTimers();
  });
  afterAll(() => {
    vi.useRealTimers();
  });

  test('Render image with error src', async () => {
    const wrapper = await mount(Image, {
      props: {
        src: imgSrc
      }
    });
    await wrapper.vm.onImgLoadError();
    expect(wrapper.find('.tu-image--error').exists()).toBe(true);
  });

  test('Footer should work', async () => {
    const wrapper = await mount(Image, {
      props: {
        src: imgSrc,
        title: 'My title'
      }
    });
    await wrapper.vm.onImgLoaded();

    expect(wrapper.find('.tu-image__footer-caption-title').text()).toBe(
      'My title'
    );
  });

  test('Should show preview on click', async () => {
    const wrapper = await mount(Image, {
      props: {
        src: imgSrc,
        renderToBody: false
      }
    });
    await wrapper.vm.onImgLoaded();
    await wrapper.find('img').trigger('click');
    expect(wrapper.emitted()).toHaveProperty('preview-visible-change');
    expect(wrapper.html()).toMatchSnapshot();
  });

  // Preview
  test('Preview should close on mask click', async () => {
    const wrapper = await getPreviewInstance();
    const maskWrapper = wrapper.find('.tu-image__preview-wrapper');
    await maskWrapper.trigger('click');
    expect(wrapper.find('.tu-image__preview-wrapper').exists()).toBe(false);
  });

  test('Preview should be able to move', async () => {
    const map: any = {};
    const _addEventListener = window.addEventListener;
    Object.defineProperty(window, 'addEventListener', {
      value: vi.fn().mockImplementation((event, cb) => {
        map[event] = cb;
      })
    });
    window.removeEventListener = vi.fn().mockImplementation((event) => {
      delete map[event];
    });
    const wrapper = await getPreviewInstance();
    const imageElement = wrapper.find('.tu-image__preview-img');
    await imageElement.trigger('mousedown');
    expect(imageElement.attributes('class')).toContain(
      'tu-image__preview-img--moving'
    );
    await map.mouseup({ preventDefault: () => {} });
    expect(imageElement.attributes('class')).not.toContain(
      'tu-image__preview-img--moving'
    );
    Object.defineProperty(window, 'addEventListener', {
      value: _addEventListener
    });
  });

  test('Preview fullscreen should work', async () => {
    const wrapper = await getPreviewInstance();
    const fullscreenAction = wrapper.findAll(
      '.tu-image__preview-toolbar-action'
    )[0];
    await fullscreenAction.trigger('click');
    expect(
      wrapper.find('.tu-image__preview-img-container').attributes('style')
    ).not.toContain('scale(1, 1)');
  });

  test('Preview rotate right should work', async () => {
    const wrapper = await getPreviewInstance();
    const rotateRightAction = wrapper.findAll(
      '.tu-image__preview-toolbar-action'
    )[1];
    await rotateRightAction.trigger('click');
    expect(
      wrapper.find('.tu-image__preview-img').attributes('style')
    ).toContain('rotate(90deg)');
  });

  test('Preview rotate left should work', async () => {
    const wrapper = await getPreviewInstance();
    const rotateLeftAction = wrapper.findAll(
      '.tu-image__preview-toolbar-action'
    )[2];
    await rotateLeftAction.trigger('click');
    expect(
      wrapper.find('.tu-image__preview-img').attributes('style')
    ).toContain('rotate(270deg)');
  });

  test('Preview zoom in should work', async () => {
    const wrapper = await getPreviewInstance();
    const zoomInAction = wrapper.findAll(
      '.tu-image__preview-toolbar-action'
    )[3];
    await zoomInAction.trigger('click');
    expect(
      wrapper.find('.tu-image__preview-img-container').attributes('style')
    ).toContain('transform: scale(1.1,1.1);');
  });


  test('Preview zoom out should work', async () => {
    const wrapper = await getPreviewInstance();
    const zoomOutAction = wrapper.findAll(
      '.tu-image__preview-toolbar-action'
    )[4];
    await zoomOutAction.trigger('click');
    expect(
      wrapper.find('.tu-image__preview-img-container').attributes('style')
    ).toContain('transform: scale(0.9,0.9);');
  });

  // ImagePreviewGroup
  test('Should handle arrow click correctly', async () => {
    const wrapper = await mount(ImagePreviewGroup, {
      props: {
        srcList: ['https://1.jpg', 'https://2.jpg', 'https://3.jpg'],
        defaultVisible: true,
        renderToBody: false,
      },
    });
    const rightIcon = wrapper.find('.tu-image__preview-arrow-right');
    await rightIcon.trigger('click');
    expect(wrapper.emitted('change')![0]).toContain(1);
    // const leftIcon = wrapper.find('.tu-image__preview-arrow-left');
    // await leftIcon.trigger('click');
    // expect(wrapper.emitted('change')![1]).toContain(0);
  });


  // Utils getScale
  test('Should get scale correctly', () => {
    expect(getScale(1.3)).toEqual(1.5);
    expect(getScale(0.79)).toEqual(0.9);
  });


});
