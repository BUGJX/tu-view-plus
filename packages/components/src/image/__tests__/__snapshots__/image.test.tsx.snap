// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image.vue > Should show preview on click 1`] = `
"<div class=\\"tu-image tu-image--with-footer-inner\\"><img class=\\"tu-image__img\\" src=\\"http://127.0.0.1:9300/tu-view-plus/logo.svg\\">
  <!--v-if-->
  <!--v-if-->
  <!--teleport start-->
  <div class=\\"tu-image__preview\\" style=\\"z-index: 1001; position: fixed; z-index: 1001; display: block;\\">
    <div class=\\"tu-image__preview-mask image-fade-enter-from image-fade-enter-active\\" style=\\"\\"></div>
    <div tabindex=\\"0\\" class=\\"tu-image__preview-wrapper\\">
      <!-- img -->
      <div class=\\"tu-image__preview-img-container\\" style=\\"transform: scale(1,1);\\"><img src=\\"http://127.0.0.1:9300/tu-view-plus/logo.svg\\" class=\\"tu-image__preview-img\\" style=\\"transform: translate(0px, 0px) rotate(0deg);\\"></div><!-- loading -->
      <div class=\\"tu-image__preview-loading\\"><i class=\\"tu-icon\\">
          <!--v-if--><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1024 1024\\">
            <path fill=\\"currentColor\\" d=\\"M512 64a32 32 0 0 1 32 32v192a32 32 0 0 1-64 0V96a32 32 0 0 1 32-32zm0 640a32 32 0 0 1 32 32v192a32 32 0 1 1-64 0V736a32 32 0 0 1 32-32zm448-192a32 32 0 0 1-32 32H736a32 32 0 1 1 0-64h192a32 32 0 0 1 32 32zm-640 0a32 32 0 0 1-32 32H96a32 32 0 0 1 0-64h192a32 32 0 0 1 32 32zM195.2 195.2a32 32 0 0 1 45.248 0L376.32 331.008a32 32 0 0 1-45.248 45.248L195.2 240.448a32 32 0 0 1 0-45.248zm452.544 452.544a32 32 0 0 1 45.248 0L828.8 783.552a32 32 0 0 1-45.248 45.248L647.744 692.992a32 32 0 0 1 0-45.248zM828.8 195.264a32 32 0 0 1 0 45.184L692.992 376.32a32 32 0 0 1-45.248-45.248l135.808-135.808a32 32 0 0 1 45.248 0zm-452.544 452.48a32 32 0 0 1 0 45.248L240.448 828.8a32 32 0 0 1-45.248-45.248l135.808-135.808a32 32 0 0 1 45.248 0z\\"></path>
          </svg>
          <!--v-if-->
        </i></div><!-- scale value -->
      <!--v-if-->
      <!-- toolbar -->
      <!--v-if-->
      <!-- close button -->
      <div class=\\"tu-image__preview-close-btn\\"><i class=\\"tu-icon\\">
          <!--v-if--><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 1024 1024\\">
            <path fill=\\"currentColor\\" d=\\"M764.288 214.592 512 466.88 259.712 214.592a31.936 31.936 0 0 0-45.12 45.12L466.752 512 214.528 764.224a31.936 31.936 0 1 0 45.12 45.184L512 557.184l252.288 252.288a31.936 31.936 0 0 0 45.12-45.12L557.12 512.064l252.288-252.352a31.936 31.936 0 1 0-45.12-45.184z\\"></path>
          </svg>
          <!--v-if-->
        </i></div><!-- group arrow -->
      <!--v-if-->
    </div>
  </div>
  <!--teleport end-->
</div>"
`;
